### PR TITLE
detect/filestore: fix options handling and impact

### PR DESCRIPTION
**Pull request by regit**
The filestore keyword had an influence on the signature matching when it should not. For example, if Suricata is analysing a traffic with a GET http request to uri /example and have the 2 following signatures loaded:

alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; sid:1; rev:1;) alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)

then the first signature will match and the second one will not.

Also the options of filestore were not honored correctly. A signature like:

alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore:to_client,tx; sid:2; rev:1;)

was not storing the file in the answer to the request.

This patch updates the logic in filestore keyword handling to fix the problems.

The patch first makes sure that a signature with filestore will hit even if there is no file in the current application layer context. Then the patch makes sure that postmatch handles the different options correctly.

As filestore keyword is not anymore preventing a match, we need to update some unit tests that were using this "feature".

Tickets: 7356 7357

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7357

Describe changes:
- rebase code
- don't filestore before tx triggering the alert
- put everything in one commit
- rework the commit message

SV_REPO=https://github.com/OISF/suricata-verify/pull/2202

### CODE REVIEW

**Comment by victorjulien** [comment]
In the commit, use
```
Ticket: 7356
Ticket: 7357
```
so our tracking can work (and the leaderboard counts them)

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 24034

**Comment by victorjulien** [comment]
SV_REPO=https://github.com/OISF/suricata-verify/pull/2202

Should be using `SV_BRANCH`

**Comment by regit** [comment]
> In the commit, use
> 
> ```
> Ticket: 7356
> Ticket: 7357
> ```
> 
> so our tracking can work (and the leaderboard counts them)

Done, thanks for the check.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 24041

**Comment by regit** [comment]
@victorjulien any update on this MR ?

**Comment by catenacyber** [comment]
I think we need to discuss the functionality/expected behavior before the implementation : see https://redmine.openinfosecfoundation.org/issues/7356#note-4

```
alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; sid:1; rev:1;) 
alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)
```
> then the first signature will match and the second one will not.

As you said (in redmine), this is because there is no files in the to_server direction

> We have no file in the direction of http.uri but from documentation filestore should not prevent the match.

I do not find such documentation.
And for me, the current behavior makes sense.

I understand the need is to store the file answered by the server if the request matches some URI...

As tried in Ticket 7357
`alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,flow; sid:2; rev:1;)`
This seems to make sense to me that it should match indeed

And bidirectional signatures should also be a cleaner way to support it in 8 cf https://github.com/OISF/suricata/pull/12405

File: src/app-layer-htp.c [hunk 1]
```diff
@@ -5809,12 +5809,12 @@ libhtp:\n\
     /* do detect */
     SigMatchSignatures(&th_v, de_ctx, det_ctx, p1);
 
-    FAIL_IF((PacketAlertCheck(p1, 1)));
+    FAIL_IF((PacketAlertCheck(p1, 2)));
```
Comment (catenacyber) [inline src/app-layer-htp.c:5812]: What does this change mean ? Can there be a comment ?

File: src/detect-engine-file.c [hunk 2]
```diff
@@ -194,7 +194,22 @@ uint8_t DetectFileInspectGeneric(DetectEngineCtx *de_ctx, DetectEngineThreadCtx
     if (ffc == NULL) {
         SCReturnInt(DETECT_ENGINE_INSPECT_SIG_CANT_MATCH_FILES);
     } else if (ffc->head == NULL) {
-        SCReturnInt(DETECT_ENGINE_INSPECT_SIG_NO_MATCH);
```
Comment (catenacyber) [inline src/detect-engine-file.c:197]: When do we get `ffc != NULL && ffc->head == NULL` ?  

This still looks fishy to me as a signature with `file.data` that does not match today because of this `SCReturnInt(DETECT_ENGINE_INSPECT_SIG_NO_MATCH)`, will cause FP with this PR...

File: src/detect-engine-file.c [hunk 3]
```diff
@@ -194,7 +194,22 @@ uint8_t DetectFileInspectGeneric(DetectEngineCtx *de_ctx, DetectEngineThreadCtx
...
+               a scope of capture where we need to prepare the capture for
+               an upcoming file. */
+            if (s->filestore_ctx && (s->filestore_ctx->scope == FILESTORE_SCOPE_TX)) {
+                /* In scope TX, we need to prepare file storage for file that could
+                   appear on the transaction so we store the transaction */
+                det_ctx->filestore[det_ctx->filestore_cnt].file_id = 0;
+                det_ctx->filestore[det_ctx->filestore_cnt].tx_id = det_ctx->tx_id;
+                det_ctx->filestore_cnt++;
+            }
+            /* Other scopes than TX are going to be handled in post match without
```
Comment (catenacyber) [inline src/detect-engine-file.c:208]: But postmatch does not make signature fail to match, does it ?

I see `(void)sigmatch_table[smd->type].Match(det_ctx, p, s, smd->ctx);` return code ignored in `DetectRunPostMatch`

File: src/detect-filestore.c [hunk 4]
```diff
@@ -98,6 +98,47 @@ void DetectFilestoreRegister(void)
...
+        uint64_t total_txs = AppLayerParserGetTxCnt(f, alstate);
+        for (uint64_t tx_id = init_tx_id; tx_id < total_txs; tx_id++) {
+            void *txv = AppLayerParserGetTx(f->proto, f->alproto, alstate, tx_id);
+            DEBUG_VALIDATE_BUG_ON(txv == NULL);
+            if (txv != NULL) {
+                AppLayerTxData *txd = AppLayerParserGetTxData(f->proto, f->alproto, txv);
+                DEBUG_VALIDATE_BUG_ON(txd == NULL);
+                if (txd != NULL) {
+                    if (toclient_dir) {
+                        txd->file_flags |= FLOWFILE_STORE_TC;
```
Comment (catenacyber) [inline src/detect-filestore.c:131]: I do not think we need to set `AppLayerTxData` `file_flags` here as we already set  `AppLayerStateData` one, and in the end we use a logical `or` of both

**Comment by regit** [comment]
> I think we need to discuss the functionality/expected behavior before the implementation : see https://redmine.openinfosecfoundation.org/issues/7356#note-4
> 
> ```
> alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; sid:1; rev:1;) 
> alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)
> ```
> 
> > then the first signature will match and the second one will not.
> 
> As you said (in redmine), this is because there is no files in the to_server direction
> 
> > We have no file in the direction of http.uri but from documentation filestore should not prevent the match.
> 
> I do not find such documentation. And for me, the current behavior makes sense.

I'm quoting the documentation here. First sentence of https://docs.suricata.io/en/latest/rules/file-keywords.html#filestore:

```
Stores files to disk if the signature matched.
```

Matched is past tense so signature is a full match. And Suricata is just going to take action on it. 


> 
> I understand the need is to store the file answered by the server if the request matches some URI...
> 
> As tried in Ticket 7357 `alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,flow; sid:2; rev:1;)` This seems to make sense to me that it should match indeed
> 
> And bidirectional signatures should also be a cleaner way to support it in 8 cf #12405

The documentation specifies the scope of extraction is a parameter. What i'm just doing here is to implement it. Only "improvisation" is the usage of first sentence of the documentation (which is clear to me) to change behavior.

File: src/detect-filestore.c [hunk 5]
```diff
@@ -207,11 +235,33 @@ static int DetectFilestorePostMatch(DetectEngineThreadCtx *det_ctx,
 {
     SCEnter();
 
+    DEBUG_VALIDATE_BUG_ON(p->flow == NULL);
+
     if (det_ctx->filestore_cnt == 0) {
```
Comment (catenacyber) [inline src/detect-filestore.c:240]: By the way, `det_ctx->filestore_cnt` looks fishy to me cf https://github.com/OISF/suricata-verify/pull/1999

Its scope is one packet when a signature can take multiple packets to match...

**Comment by catenacyber** [comment]
> I'm quoting the documentation here. First sentence of https://docs.suricata.io/en/latest/rules/file-keywords.html#filestore:
> 
> ```
> Stores files to disk if the signature matched.
> ```
> 
> Matched is past tense so signature is a full match. And Suricata is just going to take action on it.
> 

Ok, I see, my English may not be so good.

That means that `alert ip any any -> any any ( filestore; sid: 1;)` should match every packet, right ?

File: src/detect-engine-file.c (hunk 3 above)
Comment (regit) [reply src/detect-engine-file.c:208]: Yes, before we were not going to postmatch as the filestore without file was the cases where ffc or ffc->head were null and as a result it was a NO_MATCH.

**Comment by regit** [comment]
> > I'm quoting the documentation here. First sentence of https://docs.suricata.io/en/latest/rules/file-keywords.html#filestore:
> > ```
> > Stores files to disk if the signature matched.
> > ```
> > Matched is past tense so signature is a full match. And Suricata is just going to take action on it.
> 
> Ok, I see, my English may not be so good.

At least what is sure is that your accent is better than mine ;)
> 
> That means that `alert ip any any -> any any ( filestore; sid: 1;)` should match every packet, right ?

Yes, I understand it like that too.

File: src/detect-engine-file.c (hunk 2 above)
Comment (regit) [reply src/detect-engine-file.c:197]: Talking about fishy, I think it is a good catch ;) I'm going to add a suricata-verify test so we are sure this case is handled.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 24317

File: src/detect-engine-file.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-engine-file.c:197]: Ok, but do you know, in plain English, when we have the cases :
- `ffc == NULL` (protocol or direction not supporting files ?)
- `ffc != NULL && ffc->head == NULL` protocol supporting files but tx does not have a file in this direction ? (that is what I expected but this seems a wrong expectation)

File: src/detect-engine-file.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-engine-file.c:213]: I understand a bit more.
Why not handle `FILESTORE_SCOPE_TX` in postmatch as well ?

Is it not a problem if the signature matched so far but will not fully match in the end ?

**Comment by victorjulien** [comment]
We have 3 scopes for filestore:
- `file`, this is the default if the option is omitted. It means "this file".
- `tx`, all the files in the transaction.
- `flow`/`ssn`, all the files in the flow
Additionally, the direction can be specified I think.

The basic problem is that "this file" only makes sense in the context of actually inspecting a file. A rule that works on the IP layer, doesn't match a file and it has no knowledge about files.

Similarly, "this tx" only makes sense if the rule actually inspects a tx. This is similar to the tx metadata logging issue for non-tx rules. I think we should reject rules that try to specify a mismatching scope.

That leaves only the `flow`/`ssn` scope. That I think does make sense to have for rules w/o a tx or file.

Scopes:
`file` - require file inspection (file.name, file.data, etc)
`tx` - require tx inspection (basically a rule of type `app_tx`)
`flow`/`ssn` - support all

**Comment by regit** [comment]
> We have 3 scopes for filestore:
> 
> * `file`, this is the default if the option is omitted. It means "this file".
> * `tx`, all the files in the transaction.
> * `flow`/`ssn`, all the files in the flow
>   Additionally, the direction can be specified I think.
> 
> The basic problem is that "this file" only makes sense in the context of actually inspecting a file. A rule that works on the IP layer, doesn't match a file and it has no knowledge about files.
> 
> Similarly, "this tx" only makes sense if the rule actually inspects a tx. This is similar to the tx metadata logging issue for non-tx rules. I think we should reject rules that try to specify a mismatching scope.
> 
> That leaves only the `flow`/`ssn` scope. That I think does make sense to have for rules w/o a tx or file.
> 
> Scopes: `file` - require file inspection (file.name, file.data, etc) `tx` - require tx inspection (basically a rule of type `app_tx`) `flow`/`ssn` - support all

Victor, what is your conclusion here ? Not sure I understand what you try to say.

**Comment by catenacyber** [comment]
My understand of Victor's comment is that `filestore` should affect signature matching, based on its scope :
- file : only matches if there is a file (cf `FILE_SIG_NEED_FILE` ? ) , so `alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)` does not match
- tx : use `SIG_FLAG_APPLAYER` so `alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,tx; sid:2; rev:1;)` should match (I think it does not now and your PR fixes it) and also `alert ip any any -> any any ( filestore:both,tx; sid: 1;)` does not match on TCP SYN packet
- ssn/flow : use `SIG_MASK_REQUIRE_FLOW` so `alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,flow; sid:2; rev:1;)` should match (I think it does not now and your PR fixes it) and also `alert ip any any -> any any ( filestore:both,flow; sid: 1;)` does not match on ARP packets

So yeah, we should set the right signature flags based on the scope...

**Comment by victorjulien** [comment]
> My understand of Victor's comment is that `filestore` should affect signature matching, based on its scope :
> 
>     * file : only matches if there is a file (cf `FILE_SIG_NEED_FILE` ? ) , so `alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)` does not match
> 
>     * tx : use `SIG_FLAG_APPLAYER` so `alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,tx; sid:2; rev:1;)` should match (I think it does not now and your PR fixes it) and also `alert ip any any -> any any ( filestore:both,tx; sid: 1;)` does not match on TCP SYN packet
> 
>     * ssn/flow : use `SIG_MASK_REQUIRE_FLOW` so `alert http any any -> any any (msg:"exe"; http.uri; content:"exe"; filestore:both,flow; sid:2; rev:1;)` should match (I think it does not now and your PR fixes it) and also `alert ip any any -> any any ( filestore:both,flow; sid: 1;)` does not match on ARP packets
> 
> 
> So yeah, we should set the right signature flags based on the scope...

My point is we should outright fail to load the mismatches in scope. Lets make explicit which ones can be mixed, reject all other sigs.

**Comment by catenacyber** [comment]
I think the signature `alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)` is ok as it does not have mismatched scopes.
The question is about its behavior : should it match if there is no file in the http.uri direction ?

**Comment by victorjulien** [comment]
> I think the signature `alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)` is ok as it does not have mismatched scopes. The question is about its behavior : should it match if there is no file in the http.uri direction ?

It does have mismatched scopes, I think. There is no file inspection. There can be multiple files in a single http tx with multipart.

**Comment by catenacyber** [comment]
Oh, I see, thanks :-)

**Comment by regit** [comment]
> You force-pushed, do you want to do a new version of this PR ? (is it ready ?)

Error, I wanted to push a new version :/ Sorry for that. I'm going to finish it today.

**Comment by regit** [comment]
> > I think the signature `alert http any any -> any any (msg:"ex"; http.uri; content:"/example"; filestore; sid:2; rev:1;)` is ok as it does not have mismatched scopes. The question is about its behavior : should it match if there is no file in the http.uri direction ?
> 
> It does have mismatched scopes, I think. There is no file inspection. There can be multiple files in a single http tx with multipart.

I've pushed a suricata-verify test there to handle this case: https://github.com/regit/suricata-verify/commit/d3319b5b2dac9ab5191e5bd639a866cbfaa95b77

With this suricata PR (and potentially master), we end up extracting the 2 files in the POST with a simple `filestore;` keyword. I think that is what would a user expect.

It is also the best decision also in term of evasion as we are going to extract all files in context as expected. Doing just one would lead to easy evasion by posting 2 files when one is expected in the request.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .uptime | 629 | 654 | 103.97% | 

Pipeline 24355

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 24356

**Comment by catenacyber** [comment]
I think we need a decision on the SV behavior

**Comment by catenacyber** [comment]
Victor said he would take a look

**Comment by catenacyber** [comment]
Closing as stale as discussed with Eric
